File: 572. Subtree of Another Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} s
 * @param {TreeNode} t
 * @return {boolean}
 */
var isSubtree = function(s, t) {
  function traverse(temp) {
    if (!temp) return '';
    return `^${temp.val},${traverse(temp.left)},${traverse(temp.right)}$`;
  }

  const sTraversed = traverse(s);
  const tTraversed = traverse(t);
  return sTraversed.indexOf(tTraversed) > -1;
};

File: 572. Subtree of Another Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSubtree(s *TreeNode, t *TreeNode) bool {
	if t == nil {
		return true
	}

	if s == nil {
		return false
	}

	if s.Val == t.Val {
		if isIdentical(s, t) {
			return true
		}
	}

	return isSubtree(s.Left, t) || isSubtree(s.Right, t)
}

func isIdentical(s *TreeNode, t *TreeNode) bool {
	if t == nil && s == nil {
		return true
	}

	if t == nil || s == nil {
		return false
	}

	if s.Val != t.Val {
		return false
	}

	return isIdentical(s.Left, t.Left) && isIdentical(s.Right, t.Right)
}